sasmodels/models/lib/magnetic_functions.c

```diff
@@ -0,0 +1,130 @@
+//These functions are required for magnetic analysis models. They are copies 
+//from sasmodels/kernel_iq.c, to enables magnetic parameters for 1D and 2D models.
+
+
+static double langevin(
+    double x) {
+    // cotanh(x)-1\x
+
+    if (x < 0.00001) {
+        // avoid dividing by zero
+        return 1.0/3.0*x-1.0/45.0 * pow(x, 3) + 2.0/945.0 * pow(x, 5) - 1.0/4725.0 * pow(x, 7);
```

Comment:
Continued fraction has better convergence. Also note that floating point needs decimal points in order for the single/double conversion  to work properly.

```c
if (x < 0.1) {
  const double xsq = x*x
  return x / (3. + xsq / (5. + xsq/(7. + xsq/(9.))));
} else {
  return 1./tanh(x)-1./x;
}
```
and for `longevinx_x`
```c
if (x < 0.1) {
  const double xsq = x*x
  return 1./ (3. + xsq / (5. + xsq/(7. + xsq/(9.))));
} else {
  return longevin(x)/x;
}
```